src/backend/storeManagers/gog/games.ts

```diff
@@ -882,3 +894,157 @@ export function isGameAvailable(appName: string) {
+    .catch(() => {
+      return null
+    })
+}
+
+export async function updateGOGPlaytime(
+  appName: string,
+  startPlayingDate: Date,
+  finishedPlayingDate: Date
+) {
+  // Let server know about new session
+  const sessionDate = Math.floor(startPlayingDate.getTime() / 1000) // In seconds
+  const time = Math.floor(
+    (finishedPlayingDate.getTime() - startPlayingDate.getTime()) / 1000 / 60
+  ) // In minutes
+
+  // It makes no sense to post 0 minutes of playtime
+  if (time < 1) {
+    return
+  }
+
+  const data = {
+    session_date: sessionDate,
+    time
+  }
+  const userData: UserData | undefined = configStore.get_nodefault('userData')
+
+  if (!userData) {
+    logWarning(['Unable to post session, userData not present'], {
+      prefix: LogPrefix.Gog
+    })
+    return
+  }
+
+  if (!isOnline()) {
+    logWarning(['App offline, unable to post new session at this time'], {
+      prefix: LogPrefix.Gog
+    })
+    const alreadySetData = playtimeSyncQueue.get(userData.galaxyUserId, [])
+    alreadySetData.push({ ...data, appName })
+    playtimeSyncQueue.set(userData.galaxyUserId, alreadySetData)
+    runOnceWhenOnline(syncQueuedPlaytimeGOG)
+    return
+  }
+
+  const response = await postPlaytimeSession({ ...data, appName })
+
+  if (!response || response.status !== 201) {
+    logError('Failed to post session', { prefix: LogPrefix.Gog })
+    const alreadySetData = playtimeSyncQueue.get(userData.galaxyUserId, [])
+    alreadySetData.push({ ...data, appName })
+    playtimeSyncQueue.set(userData.galaxyUserId, alreadySetData)
+    return
+  }
+
+  logInfo('Posted session to gameplay.gog.com', { prefix: LogPrefix.Gog })
+}
+
+export async function syncQueuedPlaytimeGOG() {
+  if (playtimeSyncQueue.has('lock')) {
+    return
+  }
+  playtimeSyncQueue.set('lock', [])
+  const userData: UserData | undefined = configStore.get_nodefault('userData')
+  if (!userData) {
+    logError('Unable to syncQueued playtime, userData not present', {
+      prefix: LogPrefix.Gog
+    })
+    return
+  }
+  const queue = playtimeSyncQueue.get(userData.galaxyUserId, [])
+  const failed = []
+
+  for (const session of queue) {
+    if (!isOnline) {
+      failed.push(session)
+    }
+    const response = await postPlaytimeSession(session)
+
+    if (!response || response.status !== 201) {
+      logError('Failed to post session', { prefix: LogPrefix.Gog })
+      failed.push(session)
+    }
+  }
+  playtimeSyncQueue.set(userData.galaxyUserId, failed)
+  playtimeSyncQueue.delete('lock')
+  logInfo(
+    ['Finished posting sessions to gameplay.gog.com', 'failed:', failed.length],
+    {
+      prefix: LogPrefix.Gog
+    }
+  )
+}
+
+export async function getGOGPlaytime(
+  appName: string
+): Promise<number | undefined> {
+  if (!isOnline()) {
+    return
+  }
+  const credentials = await GOGUser.getCredentials()
+  const userData: UserData | undefined = configStore.get_nodefault('userData')
+
+  if (!credentials || !userData) {
+    return
+  }
```

Comment:
we do this 3 checks in a few places, maybe this can be extracted to a function like `canSync()` that either logs or maybe returns the error string, or true? so we can just do something like:

```
const canSyncResult = canSync()
if (canSyncResult !== true) {
  // log the error with added context like:
  logWarning(`Failed to get play time: ${canSyncResult}`)
  return null
}

// do the actual request
```

Reply:
We could do it that way, but don't we actually need credentials and userData later on? Loading them again from gogdl is no-op in my opinion

Reply:
yeah sorry, I missed that, maybe we can do something like the `runOnceWhenOnline` function, passing a callback

have a function like `runIfCanSync` and you pass a function that receives the userData, credentials, etc

something like:

```
const runIfCanSync = (callback) => {
  // do all the checks
  if (canSync) callback(userData,credentials,whatever)
}

// and then

runIfCanSync((userData, credentials, whatever) => {
  axios.post(`user/${userData.something}`)....
})
```

what do you think? then you can put whatever you want in that callback